src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionIterator.java

```diff
@@ -0,0 +1,226 @@
+            if ( ! readShard.getInterval().contains(pileup) ) {
+                continue;
+            }
+
+            // Pop any new pending regions off of the activity profile. These pending regions will not become ready
+            // until we've traversed all the reads that belong in them.
+            //
+            // Ordering matters here: need to check for forceConversion before adding current pileup to the activity profile
+            if ( ! activityProfile.isEmpty() ) {
+                final boolean forceConversion = pileup.getLocation().getStart() != activityProfile.getEnd() + 1;
+                pendingRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, forceConversion));
+            }
+
+            // Add the current pileup to the activity profile
+            if ( readShard.getPaddedInterval().contains(pileup.getLocation()) ) {
+                final SimpleInterval pileupInterval = new SimpleInterval(pileup);
+                final ReferenceContext pileupRefContext = new ReferenceContext(reference, pileupInterval);
+                final FeatureContext pileupFeatureContext = new FeatureContext(features, pileupInterval);
+
+                final ActivityProfileState profile = evaluator.isActive(pileup, pileupRefContext, pileupFeatureContext);
+                activityProfile.add(profile);
+            }
+
+            // A pending region only becomes ready once our locus iterator has advanced beyond the end of its extended span
+            // (this ensures that we've loaded all reads that belong in the new region)
+            if ( ! pendingRegions.isEmpty() && IntervalUtils.isAfter(pileup.getLocation(), pendingRegions.peek().getExtendedSpan(), readHeader.getSequenceDictionary()) ) {
+                nextRegion = pendingRegions.poll();
+            }
+        }
+
+        // When we run out of loci, close out the activity profile, and close out any remaining pending regions one at a time
+        if ( ! locusIterator.hasNext() ) {
+            if ( ! activityProfile.isEmpty() ) {
+                // Pop the activity profile a final time with forceConversion == true
+                pendingRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, true));
+            }
+
+            if ( ! pendingRegions.isEmpty() ) {
+                nextRegion = pendingRegions.poll();
+            }
+        }
+
+        // If there's a region ready, fill it with reads before returning
+        if ( nextRegion != null ) {
+            fillNextAssemblyRegionWithReads(nextRegion);
+        }
+
+        return nextRegion;
+    }
+
+    private void fillNextAssemblyRegionWithReads( final AssemblyRegion region ) {
+        // First we need to check the previous region for reads that also belong in this region
+        if ( previousRegion != null ) {
+            for ( final GATKRead previousRegionRead : previousRegion.getReads() ) {
```

Comment:
Are you entirely sure that previousRegion.getReads() is safe in this case? Is it possible that the reads might have been removed or filtered before the next region reaches into it?

Reply:
You're right, it might not be safe. Switching to storing the previous region's reads directly instead.